### PR TITLE
fix: bam::Record:new should return a valid record

### PR DESCRIPTION
**Pull request by nh13**
See #339.

### CODE REVIEW

**Comment by nh13** [comment]
@jch-13 would you mind approving running the workfllow?

**Comment by dcroote** [comment]
CI run approved

File: src/bam/record.rs [hunk 1]
```diff
@@ -113,12 +113,19 @@ fn extranul_from_qname(qname: &[u8]) -> usize {
 impl Record {
     /// Create an empty BAM record.
     pub fn new() -> Self {
-        Record {
+        let record = Record {
```
Comment (nh13) [inline src/bam/record.rs:116]:
```suggestion
        let mut record = Record {
```

**Comment by nh13** [comment]
All tests passing!

**Comment by dcroote** [comment]
I'm not the best to review this, but can a test be added to cover the issue this solves?

File: src/bam/record.rs [hunk 2]
```diff
@@ -113,12 +113,19 @@ fn extranul_from_qname(qname: &[u8]) -> usize {
...
             cigar: None,
             header: None,
-        }
+        };
+        // Developer note: these are needed so the returned record is properly
+        // initialized as unmapped.
+        record.set_tid(-1);
+        record.set_pos(-1);
+        record.set_mpos(-1);
+        record.set_mtid(-1);
```
Comment (nh13) [inline src/bam/record.rs:127]: I would argue that we should also set the record to be unmapped by default.  I mean, how can it be mapped if the tid/pos are -1?

File: src/bam/mod.rs [hunk 3]
```diff
@@ -2818,4 +2818,59 @@ CCCCCCCCCCCCCCCCCCC"[..],
...
+        }
+
+        // Read the record
+        {
+            // Build th reader
+            let mut reader = Reader::from_path(&bampath).expect("Error opening file.");
+
+            // Read the record
+            let mut rec = Record::new();
+            match bam.read(&mut rec) {
```
Comment (nh13) [inline src/bam/mod.rs:2863]:
```suggestion
            match reader.read(&mut rec) {
```

**Comment by nh13** [comment]
@dcroote who is a good person to tag for review?

**Comment by brainstorm** [comment]
Happy to help you out Nils! Please fix the no-default-features compile errors and the clippy warnings, the fix looks good.

File: src/bam/mod.rs [hunk 4]
```diff
@@ -2818,4 +2818,59 @@ CCCCCCCCCCCCCCCCCCC"[..],
...
+
+            // Add the version
+            header.push_record(
+                HeaderRecord::new(b"HD")
+                    .push_tag(b"VN", &"1.6")
+                    .push_tag(b"SO", &"unsorted"),
+            );
+
+            // Build the writer
+            let mut writer = Writer::from_path(bampath, &header, Format::Bam).unwrap();
```
Comment (nh13) [inline src/bam/mod.rs:2844]:
```suggestion
            let mut writer = Writer::from_path(&bampath, &header, Format::Bam).unwrap();
```
```suggestion
            let mut writer = Writer::from_path(bampath, &header, Format::Bam).unwrap();
```

File: src/bam/mod.rs [hunk 5]
```diff
@@ -2268,7 +2268,7 @@ CCCCCCCCCCCCCCCCCCC"[..],
         where
             F: Fn(&record::Record) -> Option<bool>,
         {
-            let mut bam_reader = Reader::from_path(bamfile).unwrap(); // internal functions, just unwarp
+            let mut bam_reader = Reader::from_path(bamfile).unwrap(); // internal functions, just unwrap
```
Comment (nh13) [inline src/bam/mod.rs:2271]: found a typo, fixing

File: src/bam/record.rs [hunk 6]
```diff
@@ -113,12 +113,23 @@ fn extranul_from_qname(qname: &[u8]) -> usize {
...
+        let mut record = Record {
             inner: unsafe { MaybeUninit::zeroed().assume_init() },
             own: true,
             cigar: None,
             header: None,
-        }
+        };
+        // The read/query name needs to be set as empty to properly initialize
+        // the record
+        record.set_qname(b"");
```
Comment (nh13) [inline src/bam/record.rs:124]: the writer would not write the record unless the query name is set to something (empty is ok)

File: src/bam/record.rs [hunk 7]
```diff
@@ -113,12 +113,23 @@ fn extranul_from_qname(qname: &[u8]) -> usize {
 impl Record {
     /// Create an empty BAM record.
     pub fn new() -> Self {
```
Comment (nh13) [inline src/bam/record.rs:115]: The goal in this PR is to that one does not need to call any other methods on `Record` to have a valid record.  Currently, this requires setting the read name and alignment positions (and unmapped flag if the positions are -1).  So this PR has an opinion, that a new record is unmapped with empty name/seq/qual/cigar.

Comment (brainstorm) [reply src/bam/record.rs:115]: Thanks for indirectly making me think about `Default` trait vs `new()`: https://users.rust-lang.org/t/when-to-use-default-trait/11190

File: src/bam/mod.rs (hunk 5 above)
Comment (brainstorm) [reply src/bam/mod.rs:2271]: startrek_nod

File: src/bam/record.rs (hunk 2 above)
Comment (brainstorm) [reply src/bam/record.rs:131]: Sounds reasonable but I'd quickly compare third party implementations (htslib/htsjdk/noodles) and/or spec footnotes (lots of gotchas there).

File: src/bam/record.rs (hunk 7 above)
Comment (brainstorm) [reply src/bam/record.rs:115]: Coincidental good read on Default: https://cj.rs/blog/rust-default-values-for-maintainability/

**Review by brainstorm — APPROVED** [review]

**Comment by brainstorm** [comment]
> Just ran into this issue and the above work-arounds resolved it. Any chance this makes it into the default anytime soon?

@holtjma Happy to merge the changes if you fix the current test/CI issues? Changes look reasonable, apologies for letting this sit for >2 years, mostly using noodles nowadays :-!

**Comment by holtjma** [comment]
@brainstorm Took a quick peak this morning to see if I could figure out what was going on.  The problem seems to be with the qname getting set and serde tests.  When I comment out the `record.set_qname(b"");`, all of the errors go away.  I tried digging into why, and had it print out the differences to stdout:
```
---- bam::record_serde::tests::test_serde_json stdout ----
encoded: [{"tid":0,"pos":1,"bin":4681,"mapq":1,"qname_len":4,"flag":16,"n_cigar":3,"seq_len":100,"mtid":-1,"mpos":-1,"isize":0,"data":[73,0,0,0,176,1,0,0,18,0,0,0,144,4,0,0,34,129,66,34,129,18,34,129,18,34,129,18,34,129,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,31,33,30,23,33,30,33,32,31,31,35,35,33,34,35,35,34,33,34,31,34,35,34,35,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,88,71,67,1,88,77,67,5,88,78,67,0,88,79,67,1,88,83,99,238,65,83,99,238,89,84,90,85,85,0]},{"tid":0,"pos":1,"bin":4681,"mapq":1,"qname_len":12,"flag":16,"n_cigar":3,"seq_len":100,"mtid":-1,"mpos":-1,"isize":0,"data":[73,73,46,49,52,57,55,56,51,57,50,0,176,1,0,0,18,0,0,0,144,4,0,0,34,129,66,34,129,18,34,129,18,34,129,18,34,129,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,31,33,30,23,33,30,33,32,31,31,35,35,33,34,35,35,34,33,34,31,34,35,34,35,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,88,71,67,1,88,77,67,5,88,78,67,0,88,79,67,1,88,83,99,238,65,83,99,238,89,84,90,85,85,0]},{"tid":0,"pos":1,"bin":4681,"mapq":1,"qname_len":4,"flag":16,"n_cigar":3,"seq_len":100,"mtid":-1,"mpos":-1,"isize":0,"data":[73,73,73,0,176,1,0,0,18,0,0,0,144,4,0,0,34,129,66,34,129,18,34,129,18,34,129,18,34,129,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,31,33,30,23,33,30,33,32,31,31,35,35,33,34,35,35,34,33,34,31,34,35,34,35,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,88,71,67,1,88,77,67,5,88,78,67,0,88,79,67,1,88,83,99,238,65,83,99,238,89,84,90,85,85,0]},{"tid":0,"pos":1,"bin":4681,"mapq":40,"qname_len":4,"flag":16,"n_cigar":3,"seq_len":100,"mtid":-1,"mpos":-1,"isize":0,"data":[73,86,0,0,176,1,0,0,18,0,0,0,144,4,0,0,34,129,66,34,129,18,34,129,18,34,129,18,34,129,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,31,33,30,23,33,30,33,32,31,31,35,35,33,34,35,35,34,33,34,31,34,35,34,35,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,88,71,67,1,88,77,67,5,88,78,67,0,88,79,67,1,88,83,99,238,65,83,99,238,89,84,90,85,85,0]},{"tid":0,"pos":1,"bin":4681,"mapq":1,"qname_len":4,"flag":16,"n_cigar":3,"seq_len":100,"mtid":-1,"mpos":-1,"isize":0,"data":[86,0,0,0,176,1,0,0,18,0,0,0,144,4,0,0,34,129,66,34,129,18,34,129,18,34,129,18,34,129,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,66,40,17,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,31,33,30,23,33,30,33,32,31,31,35,35,33,34,35,35,34,33,34,31,34,35,34,35,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,88,71,67,1,88,77,67,5,88,78,67,0,88,79,67,1,88,83,99,238,65,83,99,238,89,84,90,85,85,0]},{"tid":0,"pos":1,"bin":585,"mapq":1,"qname_len":4,"flag":2048,"n_cigar":3,"seq_len":100,"mtid":-1,"mpos":-1,"isize":0,"data":[86,73,0,0,176,1,0,0,2,106,24,0,144,4,0,0,18,129,20,34,129,20,34,129,20,34,129,20,34,17,136,24,36,24,136,40,65,17,17,24,129,130,65,24,136,130,129,65,17,136,136,66,17,24,136,136,130,24,17,17,136,24,36,24,136,129,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,31,33,30,23,33,30,33,32,31,31,35,35,33,34,35,35,34,33,34,31,34,35,34,35,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34,34]}]
[73] => [73, 0, 0]
[73, 73, 46, 49, 52, 57, 55, 56, 51, 57, 50] => [73, 73, 46, 49, 52, 57, 55, 56, 51, 57, 50]
[73, 73, 73] => [73, 73, 73]
[73, 86] => [73, 86, 0]
[86] => [86, 0, 0]
[86, 73] => [86, 73, 0]
thread 'bam::record_serde::tests::test_serde_json' panicked at src/bam/record_serde.rs:353:9:
assertion failed: `(left == right)`
```

It _seems_ like the qnames are getting 0-padded for some reason, but I'm not entirely sure why. Then those extra zeroes are triggering the mismatch in the tests.

I think Nils is correct in that the qname _should_ be set to the empty string, but it's not obvious to me what's going on with this padding or why that extra `set_qname` would trigger the padding to appear.  Any ideas on how to resolve this or poke at it further?

**Comment by holtjma** [comment]
@brainstorm 

Update: I found a putative patch, but don't seem to have permission to change this PR/branch (I think this is restricted from @nh13).  There's a one line addition to `fix_l_extranul` in `record_serde.rs`:

```
fn fix_l_extranul(rec: &mut Record) {
    // first, reset the number of extranuls to 0 for calling .qname(); then calculate how many we actually have
    rec.inner_mut().core.l_extranul = 0;
    let l_extranul = rec.qname().iter().rev().take_while(|x| **x == 0u8).count() as u8;
    rec.inner_mut().core.l_extranul = l_extranul;
}
```

The change resets extranul prior to calling qname, allowing it to get the correct full block with padding and then shrink it back down.

**Comment by nh13** [comment]
@holtjma this comes from my fork, do you want to make a PR to my branch or would you recommend something else?

**Comment by holtjma** [comment]
I was trying to avoid the extra PR since it’s a one-line change, but I can do that tomorrow if it’s preferred.

**Comment by nh13** [comment]
@holtjma I _think_ I did the right thing

**Pull request merged**